{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "qr9TnWBsGLtq"
      },
      "source": [
        "<img src=\"https://raw.githubusercontent.com/MasaakiU/MultiplexNanopore/master/resources/logo/SAVEMONEY_logo_with_letter_batch.png\"  height=\"autox\" width=\"600p\">\n",
        "\n",
        "*Simple Algorithm for Very Efficient Multiplexing of Oxford Nanopore Experiments for You!*\n",
        "\n",
        "# Overview of SAVEMONEY\n",
        "\n",
        "SAVEMONEY guides researchers to mix multiple plasmids for submission as a single sample to a commercial long-read sequencing service (e.g., Oxford Nanopore Technology), reducing overall sequencing costs while maintaining fidelity of sequencing results. Following is the outline of the procedure:\n",
        "\n",
        "- <a href=\"https://colab.research.google.com/github/MasaakiU/MultiplexNanopore/blob/master/colab/MultiplexNanopore.ipynb#Step1\">**Step 1. Pre-survey**</a> takes plasmid maps as inputs and provides users with optimal groupings of plasmids.\n",
        "- <a href=\"https://colab.research.google.com/github/MasaakiU/MultiplexNanopore/blob/master/colab/MultiplexNanopore.ipynb#Step2\">**Step 2. Submit samples**</a> according to the output of the pre-survey.\n",
        "- <a href=\"https://colab.research.google.com/github/MasaakiU/MultiplexNanopore/blob/master/colab/MultiplexNanopore.ipynb#Step3\">**Step 3. Post-analysis**</a> executes a computational deconvolution of the obtained results, and generates a consensus sequence for each plasmid.\n",
        "- An optional step, <a href=\"https://colab.research.google.com/github/MasaakiU/MultiplexNanopore/blob/master/colab/MultiplexNanopore.ipynb#Step4\">**Step 4. Visualization of results (optional)**</a> provides a platform for the detailed examination of the alignments and consensus generated in the post-analysis.\n",
        "\n",
        "For more information, please see [SAVEMONEY main notebook](https://colab.research.google.com/github/MasaakiU/MultiplexNanopore/blob/master/colab/MultiplexNanopore.ipynb) and [SAVEMONEY GitHub](https://github.com/MasaakiU/MultiplexNanopore).\n",
        "\n",
        "# What is SAVEMONEY BATCH?\n",
        "\n",
        "SAVEMONEY BATCH is designed to execute <a href=\"https://colab.research.google.com/github/MasaakiU/MultiplexNanopore/blob/master/colab/MultiplexNanopore.ipynb#Step3\">Step 3</a> multiple times at once. You have to provide:\n",
        "\n",
        "- `recommended_grouping.txt` file from <a href=\"https://colab.research.google.com/github/MasaakiU/MultiplexNanopore/blob/master/colab/MultiplexNanopore.ipynb#Step1\">Step 1</a>\n",
        "- plasmid maps (`*.fasta`, `*.fa`, or `*.fastq` files) used in <a href=\"https://colab.research.google.com/github/MasaakiU/MultiplexNanopore/blob/master/colab/MultiplexNanopore.ipynb#Step1\">Step 1</a>\n",
        "- raw sequencing results file(s) (`*.fastq` file) obtained after <a href=\"https://colab.research.google.com/github/MasaakiU/MultiplexNanopore/blob/master/colab/MultiplexNanopore.ipynb#Step2\">Step 2</a>\n",
        "\n",
        "SAVEMONEY BATCH will classify plasmid maps and raw sequencing results according to the `recommended_grouping.txt` file, then execute <a href=\"https://colab.research.google.com/github/MasaakiU/MultiplexNanopore/blob/master/colab/MultiplexNanopore.ipynb#Step3\">Step 3</a> sample (i.e., the mixture of plasmids) by sample.\n",
        "**It requires that samples be submitted based on the results of <a href=\"https://colab.research.google.com/github/MasaakiU/MultiplexNanopore/blob/master/colab/MultiplexNanopore.ipynb#Step1\">Step 1</a>.**\n",
        "\n",
        "---\n",
        "\n",
        "We would appreciate it if you could cite our manuscript if you use it toward any publications.<br>\n",
        "[Uematsu, M. and Baskin, J.M. Barcode-free multiplex plasmid sequencing using Bayesian analysis and nanopore sequencing. *eLife*. **2023**; 12: RP88794](https://doi.org/10.7554/eLife.88794.1)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "kzv3Xp0oTBHT"
      },
      "source": [
        "<a name=\"Execute\"></a>\n",
        "# Execute SAVEMONEY BATCH\n",
        "\n",
        "Follow steps below to execute the post-analysis. After the completion of all processes, `fastq_file_name.zip` files will appear in the `sample_data` directory for each `*.fastq` file. Right-click on the zip file to download it to your local storage. The zip file will also be automatically uploaded to your Google Drive, unless you uncheck `save_to_goole_drive` option below."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "D-WbFoT6LmOh"
      },
      "outputs": [],
      "source": [
        "#@markdown <a name=\"Step3no1\"></a>\n",
        "#@markdown ## 1. Upload files\n",
        "#@markdown The following three types of files must be uploaded to the `sample_data` directory, which is accessed by clicking on the file folder icon at left:\n",
        "#@markdown - `recommended_grouping.txt` file from <a href=\"https://colab.research.google.com/github/MasaakiU/MultiplexNanopore/blob/master/colab/MultiplexNanopore.ipynb#Step1\">Step 1</a>\n",
        "#@markdown - the plasmid maps (`*.fasta`, `*.fa`, or `*.dna`) used in <a href=\"https://colab.research.google.com/github/MasaakiU/MultiplexNanopore/blob/master/colab/MultiplexNanopore.ipynb#Step1\">Step 1</a>\n",
        "#@markdown - the raw nanopore sequencing results (`*.fastq`) obtained after <a href=\"https://colab.research.google.com/github/MasaakiU/MultiplexNanopore/blob/master/colab/MultiplexNanopore.ipynb#Step2\">Step 2</a>\n",
        "\n",
        "#@markdown Only one `*.fastq` file per sample is accepted for SAVEMONEY BATCH.\n",
        "\n",
        "#@markdown ## 2. Set combinations\n",
        "#@markdown Click this cell and hit `Runtime` -> `Run the focused cell`.\n",
        "#@markdown Dropdown menus will appear, select appropriate combinations.\n",
        "\n",
        "import re\n",
        "from IPython.display import display\n",
        "from ipywidgets import Dropdown, VBox, HBox, Layout, interactive_output, Label, HTML\n",
        "from pathlib import Path\n",
        "pwd = Path('/content/sample_data/')\n",
        "uploaded_recommended_grouping_path = pwd / \"recommended_grouping.txt\"\n",
        "uploaded_fastq_files = sorted([path for path in pwd.glob(\"*.fastq\")])\n",
        "uploaded_refseq_files = sorted([path for path in pwd.glob(\"*.*\") if path.suffix in (\".dna\", \".fa\", \".fasta\")])\n",
        "\n",
        "if not uploaded_recommended_grouping_path.exists():\n",
        "    raise Exception(\"Please upload `recommended_grouping.txt` file under the 'sample_data' directory!\")\n",
        "if not len(uploaded_fastq_files) > 0:\n",
        "    raise Exception(\"Please upload `*.fastq` files under the 'sample_data' directory!\")\n",
        "if not len(uploaded_refseq_files) > 0:\n",
        "    raise Exception(\"Please upload plasmid map files under the 'sample_data' directory!\")\n",
        "\n",
        "\n",
        "\"\"\" The script detect text with the following style in the `recommended_grouping.txt` file\n",
        "# recommended_grouping_txt(str)\n",
        "=== Group 1 ===\n",
        "abb1\tplasmid_name_1.dna\n",
        "abb2\tplasmid_name_2.dna\n",
        "abb3\tplasmid_name_3.dna\n",
        "\n",
        "=== Group 2 ===\n",
        "abb4\tplasmid_name_4.dna\n",
        "abb5\tplasmid_name_5.dna\n",
        "abb6\tplasmid_name_6.dna\n",
        "\"\"\"\n",
        "\n",
        "# read recommended_grouping_txt\n",
        "class RecommendedGrouping():\n",
        "    def __init__(self):\n",
        "        self.group_dict = {}\n",
        "        keep_record = False\n",
        "        new_group = \"\"\n",
        "        group_number = 0\n",
        "        with open(uploaded_recommended_grouping_path, \"r\") as f:\n",
        "            lines = f.readlines()\n",
        "            for l in lines + [\"=== Group 0 ===\"]:\n",
        "                if l.strip() == \"# recommended_grouping_txt(str)\":\n",
        "                    keep_record = True\n",
        "                elif keep_record:\n",
        "                    if l.startswith(\"=\"):\n",
        "                        self.group_dict[group_number] = new_group\n",
        "                        group_number = int(re.match(r\"=== Group ([0-9]+) ===\", l.strip()).group(1))\n",
        "                        new_group = Group(group_number)\n",
        "                    elif l.strip() == \"\":\n",
        "                        pass\n",
        "                    elif l.startswith(\"# \"):\n",
        "                        self.group_dict[group_number] = new_group\n",
        "                        keep_record = False\n",
        "                    else:\n",
        "                        new_group.add(*l.strip().split(\"\\t\"))\n",
        "    def assert_names_abbreviations(self):\n",
        "        names = []\n",
        "        abbreviations = []\n",
        "        for group in self.group_dict.values():\n",
        "            if group == \"\":\n",
        "                continue\n",
        "            names.extend(list(group.group_contents.values()))\n",
        "            abbreviations.extend(list(group.group_contents.keys()))\n",
        "        assert len(names) == len(set(names))\n",
        "        assert len(abbreviations) == len(set(abbreviations))\n",
        "    def __str__(self):\n",
        "        return \"\\n\".join([str(group) for group in self.group_dict.values()])\n",
        "    def __len__(self):\n",
        "        return len(self.group_dict) - 1 # 最初のグループは pseudo group である\n",
        "\n",
        "class Group():\n",
        "    def __init__(self, group_number):\n",
        "        self.group_number = group_number\n",
        "        self.group_contents = {}\n",
        "    def add(self, abbreviation, plasmid_name):\n",
        "        self.group_contents[abbreviation] = plasmid_name\n",
        "    def __str__(self):\n",
        "        return \"\\n\".join(f\"{k}\\t{v}\" for k, v in self.group_contents.items())\n",
        "\n",
        "rg = RecommendedGrouping()\n",
        "rg.assert_names_abbreviations()\n",
        "\n",
        "if len(uploaded_fastq_files) != len(rg):\n",
        "    raise Exception(\"The number of `*.fastq` files and the number of groups provided in `recommended_grouping.txt` file dies not match!\")\n",
        "\n",
        "# drop_down_input = 'Dropdown_input_1'\n",
        "child_widget_list = []\n",
        "class DDDict(dict):\n",
        "    def __init__(self):\n",
        "        self.ignore_event = False\n",
        "    def value_changed(self, change):\n",
        "        if self.ignore_event:\n",
        "            return\n",
        "        else:\n",
        "            old = change[\"old\"]\n",
        "            new = change[\"new\"]\n",
        "            owner  = change[\"owner\"]\n",
        "            detected_dd_list = self.search_dd_by_value(new)\n",
        "            assert len(detected_dd_list) == 2\n",
        "            if detected_dd_list[0] == owner:    another_dd = detected_dd_list[1]\n",
        "            else:                               another_dd = detected_dd_list[0]\n",
        "\n",
        "            # 値を変更\n",
        "            self.ignore_event = True\n",
        "            another_dd.value = old\n",
        "            self.ignore_event = False\n",
        "    def search_dd_by_value(self, value):\n",
        "        return [dd for group_number, dd in self.items() if dd.value == value]\n",
        "\n",
        "dd_dict = DDDict()\n",
        "nl = \"\\n\"\n",
        "for group_number, group in rg.group_dict.items():\n",
        "    if group_number == 0:   # 最初のグループはシュードグループ\n",
        "        continue\n",
        "    # 枠作り\n",
        "    dd = Dropdown(options=[path.name for path in uploaded_fastq_files], value=uploaded_fastq_files[group_number - 1].name)\n",
        "    hbox = HBox([Label(f\"=== Group {group_number} ===\", layout=Layout(width=\"100px\")), dd])\n",
        "    vbox = VBox([hbox, HTML(f'<div style=\"line-height: 1; padding-left: 10px; margin-bottom: 10px;\">{str(group).replace(nl, \"<br>\")}</div>')])\n",
        "    child_widget_list.append(vbox)\n",
        "    dd_dict[group_number] = dd\n",
        "    dd.observe(dd_dict.value_changed, names=[\"value\"])\n",
        "\n",
        "# display\n",
        "display(VBox(children=child_widget_list))"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@markdown ## 3. Execute\n",
        "#@markdown Click this cell and hit `Runtime` -> `Run after`. Set parameters below if necessary.\n",
        "save_to_google_drive = True #@param {type:\"boolean\"}\n",
        "if save_to_google_drive:\n",
        "    from pydrive.drive import GoogleDrive\n",
        "    from pydrive.auth import GoogleAuth\n",
        "    from google.colab import auth\n",
        "    from oauth2client.client import GoogleCredentials\n",
        "    auth.authenticate_user()\n",
        "    gauth = GoogleAuth()\n",
        "    gauth.credentials = GoogleCredentials.get_application_default()\n",
        "    drive = GoogleDrive(gauth)\n",
        "    print(\"You are logged into Google Drive and are good to go!\")\n",
        "\n",
        "# params\n",
        "gap_open_penalty = 3   #param {type:\"integer\"}\n",
        "gap_extend_penalty = 1 #param {type:\"integer\"}\n",
        "match_score = 1        #param {type:\"integer\"}\n",
        "mismatch_score = -2    #param {type:\"integer\"}\n",
        "param_dict = {i:globals()[i] for i in (\n",
        "    'gap_open_penalty',\n",
        "    'gap_extend_penalty',\n",
        "    'match_score',\n",
        "    'mismatch_score',\n",
        ")}\n",
        "score_threshold = 0.5  #@param {type:\"number\"}\n",
        "param_dict[\"score_threshold\"] = score_threshold\n",
        "\n",
        "#@markdown <ul>\n",
        "#@markdown <details><summary>Description</summary>\n",
        "#@markdown The <code>score_threshold</code> (previously konwn as <code>threshold_post</code>) is a user-defined value between (recommended range of 0.2–0.8) that represents a cutoff for short reads.\n",
        "#@markdown The default value of 0.5 provides a balance between including a higher number of reads and reads of sufficient length/quality.\n",
        "#@markdown For example, if the total number of reads is small but plasmids in the mixture were quite different from one another, the threshold can be lowered below 0.5 to increase the number of reads that are passed on to the next step of analysis.\n",
        "#@markdown Conversely, if the total number of reads is large and the plasmids in the mixture are highly similar to one another, the quality of subsequent analysis can be improved by raising the threshold above 0.5.\n",
        "#@markdown See also <a href=\"https://colab.research.google.com/github/MasaakiU/MultiplexNanopore/blob/master/colab/MultiplexNanopore.ipynb#Step3no5\">Step 3-5. Set threshold for assignment</a>, and \"Summary distribution\" and \"Summary scatter\" section of <a href=\"https://colab.research.google.com/github/MasaakiU/MultiplexNanopore/blob/master/colab/MultiplexNanopore.ipynb#InterpretationOfResults\">Interpretation of results (post-analysis)</a>.\n",
        "#@markdown\n",
        "#@markdown </details>\n",
        "#@markdown </ul>\n",
        "error_rate = 0.00001     #@param {type:\"number\"}\n",
        "del_mut_rate = error_rate / 4     # e.g. \"A -> T, C, G, del\"\n",
        "ins_rate   = 0.00001     #@param {type:\"number\"}\n",
        "window   = 160          #@param {type:\"number\"}\n",
        "param_dict[\"error_rate\"] = error_rate\n",
        "param_dict[\"del_mut_rate\"] = del_mut_rate\n",
        "param_dict[\"ins_rate\"] = ins_rate\n",
        "param_dict[\"window\"] = window\n",
        "\n",
        "#@markdown <ul>\n",
        "#@markdown <details><summary>Description (<code>error_rate</code> and <code>ins_rate</code>)</summary>\n",
        "#@markdown These variables represent the prior probability of errors that occur during the plasmid construction, such as PCR, ligation, and assembly.\n",
        "#@markdown <ul><li>The <code>error_rate</code> represents the prior probability that a base is replaced by another base including deletion.\n",
        "#@markdown For example, mutation or deletion from \"A\" to \"T\", \"C\", \"G\", or \"–\". These conversions are treated equaly, i.e., prior probability of conversion from \"A\" to \"T\" is a quarter of <code>error_rate</code>.\n",
        "#@markdown </li>\n",
        "#@markdown <li>The <code>ins_rate</code>represents the prior probability of insertion.\n",
        "#@markdown For example, the conversion of sequence \"AT\" to \"ANT\", where \"N\" represents one of the \"A\", \"T\", \"C\", and \"G.\n",
        "#@markdown </li>\n",
        "#@markdown </ul>\n",
        "#@markdown Regardless of these prior probability values, the results without considering them are always generated, which is the equivalent of using <code>error_rate</code> and <code>ins_rate</code> values of 0.8 and 0.8, respectively.\n",
        "#@markdown See also \"Consensus FASTQ files\" section in <a href=\"https://colab.research.google.com/github/MasaakiU/MultiplexNanopore/blob/master/colab/MultiplexNanopore.ipynb#InterpretationOfResultsPost\">Interpretation of results (post-analysis)</a>.\n",
        "#@markdown </details>\n",
        "#@markdown <details><summary>Description (<code>window</code>)</summary>\n",
        "#@markdown The value represents the maximum detectable length of repetitive sequences when wrong plasmid maps are provided. If you suspect that the region of 80 nt might be repeated adjascently two times, put the value of 160.\n",
        "#@markdown If the value is too large, it will take a long time to execute the process.\n",
        "#@markdown </details>\n",
        "#@markdown </ul>\n",
        "\n",
        "#@markdown <font color='red'>**All processes are executed automatically.**</font>\n",
        "\n",
        "# ファイルの組み合わせを取得\n",
        "grouping_list = [\n",
        "    [dd.value] + list(rg.group_dict[group_number].group_contents.values())\n",
        "    for group_number, dd in dd_dict.items()\n",
        "]\n",
        "\n",
        "# もう一度確認\n",
        "for grouping in grouping_list:\n",
        "    assert len(grouping) == len(set(grouping))\n",
        "\n",
        "\n",
        "print(\"installing dependencies...\")\n",
        "# import sys, os\n",
        "# save_stdout =  sys.stdout\n",
        "# sys.stdout = open(os.devnull, 'w')\n",
        "\n",
        "!pip install savemoney==0.2.16\n",
        "!wget -nc https://raw.githubusercontent.com/MasaakiU/MultiplexNanopore/master/resources/font/Arial.ttf\n",
        "!wget -nc https://raw.githubusercontent.com/MasaakiU/MultiplexNanopore/master/resources/font/Arial\\ Bold.ttf\n",
        "\n",
        "# sys.stdout.close()\n",
        "# sys.stdout = save_stdout\n",
        "print(\"installation: DONE\")\n",
        "\n",
        "import shutil\n",
        "import savemoney\n",
        "import matplotlib.font_manager as fm\n",
        "font_path = '/content/Arial.ttf'\n",
        "font_path_bold = '/content/Arial Bold.ttf'\n",
        "fm.fontManager.addfont(font_path)\n",
        "fm.fontManager.addfont(font_path_bold)\n",
        "\n",
        "# 実行ループ\n",
        "param_dict = {}\n",
        "tmp_folder_path = pwd / \"tmp_folder\"\n",
        "for grouping in grouping_list:\n",
        "    # 実行フォルダ内部のファイルをすべて削除\n",
        "    if tmp_folder_path.exists():\n",
        "        shutil.rmtree(tmp_folder_path)\n",
        "    # フォルダ作成・ファイル移動\n",
        "    tmp_folder_path.mkdir(exist_ok=True)\n",
        "    for file_name in grouping:\n",
        "        shutil.copy(pwd / file_name, tmp_folder_path / file_name)\n",
        "    # 中間結果があればそれも移動\n",
        "    ir_file_path = (pwd / grouping[0]).with_suffix(\".intermediate_results.ir\")\n",
        "    if ir_file_path.exists():\n",
        "        shutil.copy(ir_file_path, tmp_folder_path / ir_file_path.name)\n",
        "    # 実行\n",
        "    save_dir = savemoney.post_analysis(tmp_folder_path, pwd, **param_dict)\n",
        "    # compress as zip\n",
        "    import os\n",
        "    from pathlib import Path\n",
        "    import zipfile\n",
        "    zip_path = save_dir.with_suffix(\".zip\")\n",
        "    shutil.make_archive(pwd / zip_path.stem, 'zip', root_dir=save_dir)\n",
        "\n",
        "    if save_to_google_drive == True and drive:\n",
        "        uploaded = drive.CreateFile({'title': zip_path.name})\n",
        "        uploaded.SetContentFile(zip_path)\n",
        "        uploaded.Upload()\n",
        "        print(f\"Uploaded {zip_path} to Google Drive with ID {uploaded.get('id')}\")\n",
        "\n",
        "\n",
        "# 後処理\n",
        "if tmp_folder_path.exists():\n",
        "    shutil.rmtree(tmp_folder_path)\n",
        "\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "rGqp5H_hNNiG"
      },
      "execution_count": null,
      "outputs": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3.10.4 64-bit ('env_RAPID')",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "name": "python",
      "version": "3.10.4"
    },
    "vscode": {
      "interpreter": {
        "hash": "65fb75ef547f7f70fb8dbc6edf61adc9b4cedb2ccbdc90e1548a5e817107408e"
      }
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}