{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "qr9TnWBsGLtq"
      },
      "source": [
        "<img src=\"https://raw.githubusercontent.com/MasaakiU/MultiplexNanopore/master/resources/logo/SAVEMONEY_logo_with_letter.png\"  height=\"autox\" width=\"600p\">\n",
        "\n",
        "*Simple Algorithm for Very Efficient Multiplexing of Oxford Nanopore Experiments for You!*\n",
        "\n",
        "# Overview <font size=\"4\">(<a href=\"#ver0no2no11\">what's new in ver_0.2.12</a>)</font>\n",
        "\n",
        "SAVEMONEY guides researchers to mix multiple plasmids for submission as a single sample to a commercial long-read sequencing service (e.g., Oxford Nanopore Technology), reducing overall sequencing costs while maintaining fidelity of sequencing results. Following is the outline of the procedure:\n",
        "\n",
        "- <a href=\"#Step1\">**Step 1. Pre-survey**</a> takes plasmid maps as inputs and provides users with optimal groupings of plasmids.\n",
        "- <a href=\"#Step2\">**Step 2. Submit samples**</a> according to the output of the pre-survey.\n",
        "- <a href=\"#Step3\">**Step 3. Post-analysis**</a> executes a computational deconvolution of the obtained results, and generates a consensus sequence for each plasmid.\n",
        "- An optional step, <a href=\"#Step4\">**Step 4. Visualization of results (optional)**</a> provides a platform for the detailed examination of the alignments and consensus generated in the post-analysis.\n",
        "\n",
        "Detailed instructions are provided inside cells at each step. For more information, please see:\n",
        "\n",
        "- <a href=\"#QuickGuide\">**Quick guide**</a>, <a href=\"#InterpretationOfResults\">**Interpretation of results**</a>, <a href=\"#FAQ\">**Troubleshooting**</a>, <a href=\"#ReleaseNotes\">**Release notes**</a>\n",
        "- <details><summary>Graphical abstract</summary>\n",
        "<div><center>\n",
        "<img src=\"https://raw.githubusercontent.com/MasaakiU/MultiplexNanopore/master/resources/figures/Fig1_20230313_margin.png\" width=\"500\" style=\"display:inline-block; margin:auto\">\n",
        "</center></div>\n",
        "</details>\n",
        "- [SAVEMONEY GitHub](https://github.com/MasaakiU/MultiplexNanopore).\n",
        "- [Slide from Weill Institute Science Workshop, May 22, 2023](https://github.com/MasaakiU/MultiplexNanopore/blob/master/resources/slides/20230522_Weill-Institute-Science-Workshop.pdf)\n",
        "\n",
        "\n",
        "We would appreciate it if you could cite our manuscript if you use it toward any publications.<br>\n",
        "[Uematsu, M. and Baskin, J.M. Barcode-free multiplex plasmid sequencing using Bayesian analysis and nanopore sequencing. *eLife*. **2023**; 12: RP88794](https://doi.org/10.7554/eLife.88794.1)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "7wSeEei_TBHN"
      },
      "source": [
        "<a name=\"Step1\"></a>\n",
        "# Step 1. Pre-survey\n",
        "Before you submit your samples, plasmid maps can be analyzed by the pre-survey step to provide a recommendation for what combination of plasmids is safe to mix. If your plasmids are very different, i.e., completely different vector/backbone or insert, then you can skip this step and assume those plasmids are safe to mix together. Follow the steps below to execute pre-survey."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "BIoJ9pTSGCdX"
      },
      "outputs": [],
      "source": [
        "# install dependencies\n",
        "print(\"installing dependencies...\")\n",
        "# import sys, os\n",
        "# save_stdout =  sys.stdout\n",
        "# sys.stdout = open(os.devnull, 'w')\n",
        "\n",
        "!pip install savemoney==0.2.12\n",
        "!wget -nc https://raw.githubusercontent.com/MasaakiU/MultiplexNanopore/master/resources/font/Arial.ttf\n",
        "!wget -nc https://raw.githubusercontent.com/MasaakiU/MultiplexNanopore/master/resources/font/Arial\\ Bold.ttf\n",
        "!pip install pdf2image\n",
        "!apt-get install poppler-utils\n",
        "\n",
        "# sys.stdout.close()\n",
        "# sys.stdout = save_stdout\n",
        "print(\"installation: DONE\")\n",
        "\n",
        "#@markdown <a name=\"Step1no1\"></a>\n",
        "#@markdown ## Step 1-1. Upload plasmid maps\n",
        "#@markdown Click on the folder icon at left and then upload plasmid maps, i.e., expected plasmid sequences, to the `sample_data` directory.\n",
        "#@markdown Currently either `*.fasta`, `*.fa` or `*.dna` files are supported.\n",
        "#@markdown <br>\n",
        "\n",
        "#@markdown ## Step 1-2. Advanced settings\n",
        "\n",
        "gap_open_penalty = 3   #param {type:\"integer\"}\n",
        "gap_extend_penalty = 1 #param {type:\"integer\"}\n",
        "match_score = 1        #param {type:\"integer\"}\n",
        "mismatch_score = -2    #param {type:\"integer\"}\n",
        "number_of_groups = 1   #@param {type:\"number\"}\n",
        "distance_threshold = 6 #@param {type:\"number\"}\n",
        "\n",
        "#@markdown <ul><details><summary>Description (<code>number_of_groups</code>)</summary>\n",
        "#@markdown Plasmids are grouped according to the <code>number_of_groups</code> specified.\n",
        "#@markdown However, if this cannot be achieved with the given <code>distance_threshold</code> (see description below), the number of groups will be automatically increased.\n",
        "#@markdown To achieve the desired <code>number_of_groups</code>, decrease the <code>distance_threshold</code>.\n",
        "#@markdown </details></ul>\n",
        "#@markdown <ul><details><summary>Description (<code>distance_threshold</code>)</summary>\n",
        "#@markdown The <code>distance_threshold</code> (previously known as <code>threshold_pre</code>) value determines the minimum number of bases that two plasmids must differ by to be mixed together in the same sample.\n",
        "#@markdown The default value of 6 is appropriate for general use but it can be\n",
        "#@markdown <ul>\n",
        "#@markdown <li>increased (quality-oriented decision) or</li>\n",
        "#@markdown <li>decreased to as low as 2 (cost-oriented decision).</li>\n",
        "#@markdown </ul>\n",
        "#@markdown See <a href=\"https://doi.org/10.7554/eLife.88794.1\">our publication</a> for more details.\n",
        "#@markdown </details></ul><br>\n",
        "\n",
        "#@markdown ## Step 1-3. Click this cell and hit `Runtime` -> `Run the focused cell`\n",
        "#@markdown <ul><details><summary>Tips</summary>\n",
        "#@markdown If you want to perform the pre-survey again with different settings but with the same plasmid set,\n",
        "#@markdown moving the <code>recommended_grouping.txt</code> output file to the <code>sample_data</code> directory will reduce the calculation time.\n",
        "#@markdown </details></ul><br>\n",
        "\n",
        "#@markdown ## Step 1-4. View pre-survey results\n",
        "#@markdown The results are displayed below, and can also be downloaded as a `recommended_grouping.zip` file from the `sample_data` directory.\n",
        "#@markdown See <a href=\"#InterpretationOfResults\">Interpretation of results</a> for details.\n",
        "\n",
        "# params\n",
        "param_dict = {i:globals()[i] for i in (\n",
        "    'gap_open_penalty',\n",
        "    'gap_extend_penalty',\n",
        "    'match_score',\n",
        "    'mismatch_score',\n",
        "    'number_of_groups',\n",
        "    'distance_threshold',\n",
        ")}\n",
        "\n",
        "# prepare\n",
        "import savemoney\n",
        "import matplotlib.font_manager as fm\n",
        "font_path = '/content/Arial.ttf'\n",
        "font_path_bold = '/content/Arial Bold.ttf'\n",
        "fm.fontManager.addfont(font_path)\n",
        "fm.fontManager.addfont(font_path_bold)\n",
        "\n",
        "# execute\n",
        "save_dir = savemoney.pre_survey('/content/sample_data/', '/content/sample_data/', **param_dict)\n",
        "\n",
        "# display\n",
        "from pdf2image import convert_from_path\n",
        "from IPython.display import display, Image\n",
        "images = convert_from_path((save_dir / savemoney.pre_survey_core.RecommendedGrouping.file_name).with_suffix(\".pdf\"))\n",
        "for image in images:\n",
        "    display(image)\n",
        "\n",
        "# compress as zip\n",
        "import os\n",
        "from pathlib import Path\n",
        "import zipfile\n",
        "zip_path = save_dir.with_suffix(\".zip\")\n",
        "os.chdir(save_dir)\n",
        "with zipfile.ZipFile(zip_path.as_posix(), 'w') as f:\n",
        "    f.write(Path(savemoney.pre_survey_core.RecommendedGrouping.file_name))\n",
        "    f.write(Path(savemoney.pre_survey_core.RecommendedGrouping.file_name).with_suffix(\".pdf\"))\n",
        "os.chdir(\"/content\")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "oeFKhV7ATBHS"
      },
      "source": [
        "<a name=\"Step2\"></a>\n",
        "# Step 2. Submit samples\n",
        "Mix plasmids according to the <a href=\"#Step1\">**Step 1. Pre-survey**</a> results and submit sample(s) for nanopore sequencing.\n",
        "<ul>\n",
        "<details><summary>Verify plasmids</summary><hr>\n",
        "   It is strongly recommended that you verify that your plasmids are built correctly by using a diagnostic restriction digestion or another method with an equivalent degree of confidence. SAVEMONEY cannot detect sequences that differ significantly from the uploaded plasmid maps.\n",
        "   <hr></details>\n",
        "\n",
        "<details><summary>Prepare samples</summary><hr>\n",
        "   Mix together plasmids according to the grouping of the pre-survey step. It is important to use the same amount of each plasmid in the mixture. For example, the Plasmidsaurus standard service currently requires samples of 10 µL at 30 ng/µL (i.e., 300 ng of total plasmid). Therefore, if the pre-survey instructs you to mix three plasmids in a single sample, provide 100 ng of each plasmid in 10 µL total of water.\n",
        "   <hr></details>\n",
        "\n",
        "<details><summary>Submit samples</summary><hr>\n",
        "Submit samples for sequencing to a third-party whole-plasmid sequencing service and obtain raw <code>*.fastq</code> files. Note that the consensus sequence provided by the commercial service is not likely to make any sense because you have provided them with a mixture of different plasmids.\n",
        "<hr></details>\n",
        "</ul>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "kzv3Xp0oTBHT"
      },
      "source": [
        "<a name=\"Step3\"></a>\n",
        "# Step 3. Post-analysis\n",
        "\n",
        "This step must be performed independently for each experimental sample mixture. Follow <a href=\"#Step3no1\">**Step 3-1** and **Step 3-2**</a> below to execute the post-analysis. All processes are executed automatically.\n",
        "\n",
        "After the completion of all processes, a `fastq_file_name.zip` file will appear in the `sample_data` directory. Right-click on the zip file to download it to your local storage. The zip file will also be automatically uploaded to your Google Drive, unless you uncheck `save_to_goole_drive` option at <a href=\"#Step3no4\">**Step 3-4**</a> before execution."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "cellView": "form",
        "id": "D-WbFoT6LmOh"
      },
      "outputs": [],
      "source": [
        "#@markdown <a name=\"Step3no1\"></a>\n",
        "#@markdown ## Step 3-1. Upload files\n",
        "#@markdown The following two types of files must be uploaded to the `sample_data` directory, which is accessed by clicking on the file folder icon at left:\n",
        "#@markdown - the plasmid map for each plasmid within the sample mixture (`*.fasta`, `*.fa`, or `*.dna`)\n",
        "#@markdown - the raw nanopore sequencing results (`*.fastq`)\n",
        "\n",
        "#@markdown If your results are split into multiple `*.fastq` files, just upload all of them;\n",
        "#@markdown SAVEMONEY will merge those files and execute the analysis.\n",
        "\n",
        "#@markdown ## Step 3-2. Click this cell and hit `Runtime` -> `Run after`\n",
        "\n",
        "#@markdown <font color='red'>**All processes below are executed automatically.**</font>\n",
        "\n",
        "#@markdown Post-analysis is a multi-step process, and the cell that is currently running will be indicated by a circle with a stop sign next to it.\n",
        "#@markdown There are some user-modifiable variables. If you would like to change them, please refer to the instructions for each cell."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "5TrFiqLrx8Mq"
      },
      "outputs": [],
      "source": [
        "from pathlib import Path\n",
        "pwd = Path('/content/sample_data/')\n",
        "uploaded_fastq_files = [path for path in pwd.glob(\"*.fastq\")]\n",
        "uploaded_refseq_files = [path for path in pwd.glob(\"*.*\") if path.suffix in (\".dna\", \".fa\", \".fasta\")]\n",
        "\n",
        "if not len(uploaded_fastq_files) > 0:\n",
        "    raise Exception(\"Please upload fastq files under the 'sample_data' directory!\")\n",
        "if not len(uploaded_refseq_files) > 0:\n",
        "    raise Exception(\"Please upload reference files under the 'sample_data' directory!\")\n",
        "\n",
        "#@markdown ## Step 3-3. Select files\n",
        "#@markdown <ul><details><summary>Tips</summary>\n",
        "#@markdown If you do not want to use all files you uploaded, you can select which files to use at this step.\n",
        "#@markdown Click this cell and hit <code>Runtime</code> -> <code>Run the focused cell</code>, then checkboxes will appear.\n",
        "#@markdown Select files you want to include, and execute the following steps by clicking next cell <a href=\"#Step3no4\"><b>Step 3-4. Execute alignment</b></a> and hitting <code>Runtime</code> -> <code>Run after</code>.\n",
        "#@markdown </details></ul>\n",
        "\n",
        "from IPython.display import display\n",
        "from ipywidgets import Checkbox, VBox, Layout, interactive_output, Label\n",
        "# widgets\n",
        "child_widget_list = [Label(\"# fastq files\")]\n",
        "arg_dict = {}\n",
        "for uploaded_fastq in uploaded_fastq_files:\n",
        "    ckbx = Checkbox(value=True, description=uploaded_fastq.name, indent=False, layout=Layout(width='80%'))\n",
        "    child_widget_list.append(ckbx)\n",
        "    arg_dict[uploaded_fastq.name] = ckbx\n",
        "child_widget_list.extend([Label(\" \"), Label(\"# dna files\")])\n",
        "for uploaded_refseq in uploaded_refseq_files:\n",
        "    ckbx = Checkbox(value=True, description=uploaded_refseq.name, indent=False, layout=Layout(width='80%'))\n",
        "    child_widget_list.append(ckbx)\n",
        "    arg_dict[uploaded_refseq.name] = ckbx\n",
        "ui = VBox(children=child_widget_list)\n",
        "\n",
        "# observation function\n",
        "def select_data(**kwargs):\n",
        "    N_fastq = sum([1 for k, ckbx in kwargs.items() if (ckbx & k.endswith(\".fastq\"))])\n",
        "    N_refseq = sum([1 for k, ckbx in kwargs.items() if ckbx & (not k.endswith(\".fastq\"))])\n",
        "    if (N_fastq > 0) & (N_refseq > 0):\n",
        "        print(f\"\\n{N_fastq} fastq files selected\\n{N_refseq} reference sequence files selected\")\n",
        "    else:\n",
        "        error_text = \"\"\n",
        "        if N_fastq == 0:\n",
        "            error_text += f\"\\nPlease select at least 1 fastq file!\"\n",
        "        if N_refseq == 0:\n",
        "            error_text += f\"\\nPlease select at least 1 reference sequence file!\"\n",
        "        print(error_text)\n",
        "\n",
        "# display\n",
        "out = interactive_output(select_data, arg_dict)\n",
        "display(ui, out)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "lp4ACHXs8THv"
      },
      "outputs": [],
      "source": [
        "#@markdown <a name=\"Step3no4\"></a>\n",
        "#@markdown ## Step 3-4. Execute alignment\n",
        "save_to_google_drive = True #@param {type:\"boolean\"}\n",
        "if save_to_google_drive:\n",
        "    from pydrive.drive import GoogleDrive\n",
        "    from pydrive.auth import GoogleAuth\n",
        "    from google.colab import auth\n",
        "    from oauth2client.client import GoogleCredentials\n",
        "    auth.authenticate_user()\n",
        "    gauth = GoogleAuth()\n",
        "    gauth.credentials = GoogleCredentials.get_application_default()\n",
        "    drive = GoogleDrive(gauth)\n",
        "    print(\"You are logged into Google Drive and are good to go!\")\n",
        "\n",
        "# params\n",
        "gap_open_penalty = 3   #param {type:\"integer\"}\n",
        "gap_extend_penalty = 1 #param {type:\"integer\"}\n",
        "match_score = 1        #param {type:\"integer\"}\n",
        "mismatch_score = -2    #param {type:\"integer\"}\n",
        "param_dict = {i:globals()[i] for i in (\n",
        "    'gap_open_penalty',\n",
        "    'gap_extend_penalty',\n",
        "    'match_score',\n",
        "    'mismatch_score',\n",
        ")}\n",
        "\n",
        "print(\"installing dependencies...\")\n",
        "# import sys, os\n",
        "# save_stdout =  sys.stdout\n",
        "# sys.stdout = open(os.devnull, 'w')\n",
        "\n",
        "!pip install savemoney==0.2.12\n",
        "!wget -nc https://raw.githubusercontent.com/MasaakiU/MultiplexNanopore/master/resources/font/Arial.ttf\n",
        "!wget -nc https://raw.githubusercontent.com/MasaakiU/MultiplexNanopore/master/resources/font/Arial\\ Bold.ttf\n",
        "\n",
        "# sys.stdout.close()\n",
        "# sys.stdout = save_stdout\n",
        "print(\"installation: DONE\")\n",
        "\n",
        "# get files selected above\n",
        "fastq_paths = []\n",
        "plasmid_map_paths = []\n",
        "for child_widget in child_widget_list:\n",
        "    if not isinstance(child_widget, Checkbox):\n",
        "        continue\n",
        "    if child_widget.value & child_widget.description.endswith(\".fastq\"):\n",
        "        fastq_paths.append(pwd / child_widget.description)\n",
        "    elif child_widget.value & (not child_widget.description.endswith(\".fastq\")):\n",
        "        plasmid_map_paths.append(pwd / child_widget.description)\n",
        "\n",
        "# prepare\n",
        "import matplotlib.font_manager as fm\n",
        "font_path = '/content/Arial.ttf'\n",
        "font_path_bold = '/content/Arial Bold.ttf'\n",
        "fm.fontManager.addfont(font_path)\n",
        "fm.fontManager.addfont(font_path_bold)\n",
        "\n",
        "# execute\n",
        "from savemoney import post_analysis_core as pac\n",
        "from savemoney.modules import my_classes as mc\n",
        "from savemoney.post_analysis import default_post_analysis_param_dict\n",
        "param_dict = {key: param_dict.get(key, val) for key, val in default_post_analysis_param_dict.items()}\n",
        "mc.assert_param_dict(param_dict)\n",
        "if len(plasmid_map_paths) == 0:\n",
        "    raise Exception(f\"Error: No plasmid map file was detected!\")\n",
        "if len(fastq_paths) == 0:\n",
        "    raise Exception(f\"Error: No fastq file was detected!\")\n",
        "\n",
        "# 1. Prepare objects\n",
        "ref_seq_list = [mc.MyRefSeq(plasmid_map_path) for plasmid_map_path in plasmid_map_paths]\n",
        "my_fastq = mc.MyFastQ.combine([mc.MyFastQ(fastq_path) for fastq_path in fastq_paths])\n",
        "save_dir = mc.new_dir_path_wo_overlap(Path('/content/sample_data/') / my_fastq.combined_name_stem, spacing=\"_\")\n",
        "save_dir.mkdir()\n",
        "\n",
        "# 2. Execute alignment: load if any previous score_matrix if possible\n",
        "result_dict, ref_seq_list = pac.execute_alignment(ref_seq_list, my_fastq, param_dict, save_dir)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "7TNn4euoS-ej"
      },
      "outputs": [],
      "source": [
        "#@markdown <a name=\"Step3no5\"></a>\n",
        "#@markdown ## Step 3-5. Set threshold for assignment\n",
        "\n",
        "score_threshold = 0.5  #@param {type:\"number\"}\n",
        "param_dict[\"score_threshold\"] = score_threshold\n",
        "\n",
        "#@markdown <ul>\n",
        "#@markdown <details><summary>Description</summary>\n",
        "#@markdown The <code>score_threshold</code> (previously konwn as <code>threshold_post</code>) is a user-defined value between (recommended range of 0.2–0.8) that represents a cutoff for short reads.\n",
        "#@markdown The default value of 0.5 provides a balance between including a higher number of reads and reads of sufficient length/quality.\n",
        "#@markdown For example, if the total number of reads is small but plasmids in the mixture were quite different from one another, the threshold can be lowered below 0.5 to increase the number of reads that are passed on to the next step of analysis.\n",
        "#@markdown Conversely, if the total number of reads is large and the plasmids in the mixture are highly similar to one another, the quality of subsequent analysis can be improved by raising the threshold above 0.5.\n",
        "#@markdown See also \"Summary distribution\" and \"Summary scatter\" section of <a href=\"#InterpretationOfResults\">Interpretation of results (post-analysis)</a>.\n",
        "#@markdown </details>\n",
        "#@markdown <details><summary>Tips</summary>\n",
        "#@markdown <ul><li>\n",
        "#@markdown You can change the <code>score_threshold</code> value even after all processes are finished. Just change the value and press <code>Runtime</code> -> <code>Run after</code> (i.e., run Step 3-5 and after).\n",
        "#@markdown You do not have to execute the time-consuming <a href=\"#Step3no4\"><b>Step 3-4. Execute alignment</b></a> again.\n",
        "#@markdown </li><li>\n",
        "#@markdown If you want to re-analyze the data after a while, i.e., after the GoogleColab session is over,\n",
        "#@markdown upload the <code>*.intermediate_results.ir</code> file together with other files at <a href=\"#Step3no1\"><b>Step 3-1. Upload files</b></a> and run <a href=\"#Step3no1\"><b>Step 3-2. Click this cell...</b></a> and after.\n",
        "#@markdown See also \"Other files\" section of <a href=\"#InterpretationOfResults\">Interpretation of results (post-analysis)</a>.\n",
        "#@markdown </li></ul>\n",
        "#@markdown </details>\n",
        "#@markdown </ul>\n",
        "\n",
        "# prepare\n",
        "if (save_dir / my_fastq.combined_name_stem).with_suffix(\".summary_scores.csv\").exists():\n",
        "    original_save_dir = save_dir\n",
        "    save_dir = mc.new_dir_path_wo_overlap(Path('/content/sample_data/') / my_fastq.combined_name_stem, spacing=\"_\")\n",
        "    save_dir.mkdir()\n",
        "    for suffix in [\".intermediate_results.ir\"]:\n",
        "        file_name = my_fastq.combined_name_stem + suffix\n",
        "        original_file_path = original_save_dir / file_name\n",
        "        if original_file_path.exists():\n",
        "            import shutil\n",
        "            shutil.copy(original_file_path, save_dir / file_name)\n",
        "\n",
        "# 3. normalize alignment score and set threshold for assignment\n",
        "query_assignment = pac.normalize_scores_and_apply_threshold(ref_seq_list, my_fastq, result_dict, param_dict)\n",
        "\n",
        "pac.draw_and_save_query_assignment(query_assignment, save_dir, display_plot=False)\n",
        "\n",
        "# display\n",
        "from IPython.display import SVG, display\n",
        "display(SVG(filename=Path(save_dir) / f\"{my_fastq.combined_name_stem}.summary_distribution.svg\"))\n",
        "display(SVG(filename=Path(save_dir) / f\"{my_fastq.combined_name_stem}.summary_scatter.svg\"))\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "zv-S4MbEgnGN"
      },
      "outputs": [],
      "source": [
        "#@markdown <a name=\"Step3no6\"></a>\n",
        "#@markdown ## Step 3-6. Calculate consensus\n",
        "\n",
        "error_rate = 0.00001     #@param {type:\"number\"}\n",
        "del_mut_rate = error_rate / 4     # e.g. \"A -> T, C, G, del\"\n",
        "ins_rate   = 0.00001     #@param {type:\"number\"}\n",
        "window   = 160          #@param {type:\"number\"}\n",
        "param_dict[\"error_rate\"] = error_rate\n",
        "param_dict[\"del_mut_rate\"] = del_mut_rate\n",
        "param_dict[\"ins_rate\"] = ins_rate\n",
        "param_dict[\"window\"] = window\n",
        "\n",
        "#@markdown <ul>\n",
        "#@markdown <details><summary>Description (<code>error_rate</code> and <code>ins_rate</code>)</summary>\n",
        "#@markdown These variables represent the prior probability of errors that occur during the plasmid construction, such as PCR, ligation, and assembly.\n",
        "#@markdown <ul><li>The <code>error_rate</code> represents the prior probability that a base is replaced by another base including deletion.\n",
        "#@markdown For example, mutation or deletion from \"A\" to \"T\", \"C\", \"G\", or \"–\". These conversions are treated equaly, i.e., prior probability of conversion from \"A\" to \"T\" is a quarter of <code>error_rate</code>.\n",
        "#@markdown </li>\n",
        "#@markdown <li>The <code>ins_rate</code>represents the prior probability of insertion.\n",
        "#@markdown For example, the conversion of sequence \"AT\" to \"ANT\", where \"N\" represents one of the \"A\", \"T\", \"C\", and \"G.\n",
        "#@markdown </li>\n",
        "#@markdown </ul>\n",
        "#@markdown Regardless of these prior probability values, the results without considering them are always generated, which is the equivalent of using <code>error_rate</code> and <code>ins_rate</code> values of 0.8 and 0.8, respectively.\n",
        "#@markdown See also \"Consensus FASTQ files\" section in <a href=\"#InterpretationOfResultsPost\">Interpretation of results (post-analysis)</a>.\n",
        "#@markdown </details>\n",
        "#@markdown <details><summary>Description (<code>window</code>)</summary>\n",
        "#@markdown The value represents the maximum detectable length of repetitive sequences when wrong plasmid maps are provided. If you suspect that the region of 80 nt might be repeated adjascently two times, put the value of 160.\n",
        "#@markdown If the value is too large, it will take a long time to execute the process.\n",
        "#@markdown </details>\n",
        "#@markdown </ul>\n",
        "\n",
        "# prepare\n",
        "if (save_dir / my_fastq.combined_name_stem).with_suffix(\".log\").exists():\n",
        "    original_save_dir = save_dir\n",
        "    save_dir = mc.new_dir_path_wo_overlap(Path('/content/sample_data/') / my_fastq.combined_name_stem, spacing=\"_\")\n",
        "    save_dir.mkdir()\n",
        "    for suffix in [\n",
        "        \".intermediate_results.ir\",\n",
        "        \".summary_distribution.svg\",\n",
        "        \".summary_rotated.svg\",\n",
        "        \".summary_scatter.svg\",\n",
        "        \".summary_scores.csv\",\n",
        "    ]:\n",
        "        file_name = my_fastq.combined_name_stem + suffix\n",
        "        original_file_path = original_save_dir / file_name\n",
        "        if original_file_path.exists():\n",
        "            import shutil\n",
        "            shutil.copy(original_file_path, save_dir / file_name)\n",
        "\n",
        "# 4. MSA/consensus\n",
        "my_msa_list = pac.execute_msa(result_dict, query_assignment, param_dict)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "tcTfIX4tC-R9"
      },
      "outputs": [],
      "source": [
        "#@markdown <a name=\"Step3no7\"></a>\n",
        "#@markdown ## Step 3-7. Export results\n",
        "#@markdown See <a href=\"#InterpretationOfResultsPost\">Interpretation of results (post-analysis)</a> for the detailed description of result files.\n",
        "\n",
        "# 5. EXPORT\n",
        "pac.export_results(my_msa_list, save_dir)\n",
        "pac.export_log(ref_seq_list, my_fastq, param_dict, query_assignment, save_dir)\n",
        "\n",
        "# compress as zip\n",
        "import os\n",
        "from pathlib import Path\n",
        "import zipfile\n",
        "zip_path = save_dir.with_suffix(\".zip\")\n",
        "os.chdir(save_dir)\n",
        "with zipfile.ZipFile(zip_path.as_posix(), 'w') as f:\n",
        "    f.write(Path(my_fastq.combined_name_stem).with_suffix(\".intermediate_results.ir\"))\n",
        "    f.write(Path(my_fastq.combined_name_stem).with_suffix(\".log\"))\n",
        "    f.write(Path(my_fastq.combined_name_stem).with_suffix(\".summary_distribution.svg\"))\n",
        "    f.write(Path(my_fastq.combined_name_stem).with_suffix(\".summary_rotated.svg\"))\n",
        "    f.write(Path(my_fastq.combined_name_stem).with_suffix(\".summary_scatter.svg\"))\n",
        "    f.write(Path(my_fastq.combined_name_stem).with_suffix(\".summary_scores.csv\"))\n",
        "    for ref_seq in ref_seq_list:\n",
        "        f.write(Path(ref_seq.path.name).as_posix() + \".ca\")\n",
        "        f.write(Path(ref_seq.path.name).as_posix() + \".consensus_with_prior.fastq\")\n",
        "        f.write(Path(ref_seq.path.name).as_posix() + \".consensus_without_prior.fastq\")\n",
        "        f.write(Path(ref_seq.path.name).as_posix() + \".gif\")\n",
        "os.chdir(\"/content\")\n",
        "\n",
        "if save_to_google_drive == True and drive:\n",
        "  uploaded = drive.CreateFile({'title': zip_path.name})\n",
        "  uploaded.SetContentFile(zip_path)\n",
        "  uploaded.Upload()\n",
        "  print(f\"Uploaded {zip_path} to Google Drive with ID {uploaded.get('id')}\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "WcA9d38GTBHc"
      },
      "source": [
        "<a name=\"Step4\"></a>\n",
        "# Step 4. Visualization of results (optional)\n",
        "Please refer to the following sections to see in which cases this cell must be executed.\n",
        "- \"Alignment consensus (<code>*.ca</code>) files\" section in <a href=\"#InterpretationOfResultsPost\">Interpretation of results (post-analysis)</a>\n",
        "- <a href=\"#FAQnoFastqDoesNotMatch\">Troubleshooting – The consensus sequences of the two FASTQ files do not match</a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "vn1KKMsNmfxt"
      },
      "outputs": [],
      "source": [
        "#@markdown ## Step 4-1. Upload files\n",
        "#@markdown Upload the `plasmid_name.ca` file you obtained after <a href=\"#Step3\">**Step 3. Post-analysis**</a> in the `sample_data` directory.<br><br>\n",
        "\n",
        "#@markdown ## Step 4-2. Set parameters\n",
        "\n",
        "target_file_path = \"plasmid_name.fa.ca\" #@param {type:\"string\"}\n",
        "target_position = 1000  #@param {type:\"number\"}\n",
        "display_range = 50      #@param {type:\"number\"}\n",
        "offset = 0              #@param {type:\"number\"}\n",
        "\n",
        "#@markdown <ul>\n",
        "#@markdown <details><summary>Description</summary>\n",
        "#@markdown Example of results executing this cell with the following parameters are displayed below:\n",
        "#@markdown <ul>\n",
        "#@markdown <li>target_position = 1120\n",
        "#@markdown <li>display_range = 50\n",
        "#@markdown <li>offset = 4\n",
        "#@markdown </ul>\n",
        "#@markdown <img src=\"https://raw.githubusercontent.com/MasaakiU/MultiplexNanopore/master/resources/figures/view_alignment_results.png\"  height=\"autox\" width=\"800p\">\n",
        "#@markdown </details>\n",
        "#@markdown </ul>\n",
        "#@markdown <br>\n",
        "\n",
        "#@markdown ## Step 4-3. Select this cell and hit `Runtime` -> `Run after`\n",
        "\n",
        "from pathlib import Path\n",
        "pwd = Path('/content/sample_data/')\n",
        "file_path = pwd / target_file_path\n",
        "if file_path.exists():\n",
        "    !pip install savemoney==0.2.12\n",
        "    from savemoney import show_consensus\n",
        "    show_consensus(\n",
        "        consensus_alignment_path=file_path,\n",
        "        center = target_position,\n",
        "        seq_range = display_range,\n",
        "        offset = offset\n",
        "    )\n",
        "else:\n",
        "    print(f\"file does not exist:\\n{file_path}\")\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Y-ikqnF_TBHd"
      },
      "source": [
        "<a name=\"QuickGuide\"></a>\n",
        "# Quick guide\n",
        "\n",
        "The algorithm permits mixing of six (or potentially even more) plasmids for sequencing with Oxford Nanopore Technology (e.g., Plasmidsaurus services) and permits mixing of plasmids with as few as two base differences.\n",
        "\n",
        "## Quick start (pre-survey)\n",
        "1. Upload plasmid maps (`*.fasta`, `*.fa`, or `*.fastq` files)\n",
        "2. Click the cell <a href=\"#Step1no1\">**Step 1-1 ~ Step 1-4**</a> and hit `Runtime` -> `Run the focused cell`\n",
        "3. Results are displayed below the cell <a href=\"#Step1no1\">**Step 1-1 ~ Step 1-4.**</a>\n",
        "\n",
        "## Quick start (post-analysis)\n",
        "1. Upload plasmid maps (`*.fasta`, `*.fa`, or `*.fastq` files)\n",
        "2. Upload raw sequencing results file(s) (`*.fastq` file)\n",
        "3. Click the cell <a href=\"#Step3no1\">**Step 3-1 ~ Step 3-2**</a> and hit `Runtime` -> `Run after`\n",
        "4. The `fastq_file_name.zip` file can be downloaded from your Google Drive or `sample_data` directory."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "jjKbsIVbTBHd"
      },
      "source": [
        "<a name=\"InterpretationOfResults\"></a>\n",
        "# Interpretation of results\n",
        "\n",
        "## Interpretation of results (pre-survey)\n",
        "Two files are generated in the `sample_data` directory.\n",
        "<ul>\n",
        "<details><summary>Heatmaps (<code>recommended_grouping.pdf</code>)</summary><hr>\n",
        "A heatmap is generated for each recommended grouping of plasmids to be mixed together. Below is an example of a pre-survey of 14 different plasmids that were classified into two groups.\n",
        "<p align=\"center\"><img src=\"https://github.com/MasaakiU/MultiplexNanopore/raw/master/resources/figures/example_recommended_group.png\" width=\"800\"/></p>\n",
        "Plasmids were classified into two groups (Group 1: P5, P9, P6, P1, P7, P11, P13; Group 2: P12, P14, P3, P2, P4, P8, P10), indicating that they can be submitted to sequencing as two samples. The values written in each panel indicate the number of bases that differ between each plasmid pair. Some of these values are emphasized by the colored frames within each group.\n",
        "\n",
        "The list of all plasmids in that group is shown. Underneath it is the values of parameters that are applied during the analysis. Note that these values do not necessarily match the values provided by the user. See\n",
        "<a href=\"#Step1\"><b>Step 1-2. Advanced settings</b></a> for details.\n",
        "\n",
        "<hr></details>\n",
        "<details><summary>Text file (<code>recommended_grouping.txt</code>)</summary><hr>\n",
        "A text file containing information displayed in the heatmap is also provided. Below is an example of a part of the output text file that shows information equivalent to the above heatmaps.\n",
        "<pre><code><font size=\"2\">&#35; recommended_grouping_txt(str)\n",
        "=== Group 1 ===\n",
        "P5  plasmid_name_5.fa\n",
        "P9  plasmid_name_9.fa\n",
        "P6  plasmid_name_6.fa\n",
        "P1  plasmid_name_1.fa\n",
        "P7  plasmid_name_7.fa\n",
        "P11 plasmid_name_11.fa\n",
        "P13 plasmid_name_13.fa<br>\n",
        "=== Group 2 ===\n",
        "P12 plasmid_name_12.fa\n",
        "P14 plasmid_name_14.fa\n",
        "P3  plasmid_name_3.fa\n",
        "P2  plasmid_name_2.fa\n",
        "P4  plasmid_name_4.fa\n",
        "P8  plasmid_name_8.fa\n",
        "P10 plasmid_name_10.fa</code></pre><hr></details>\n",
        "</ul>\n",
        "\n",
        "<a name=\"InterpretationOfResultsPost\"></a>\n",
        "## Interpretation of results (post-analysis)\n",
        "\n",
        "Several files are generated by the post-analysis. We recommend to always review \"Summary GIF\", \"Consensus FASTQ\", \"Summary distribution\", and \"Summary scatter\" files.\n",
        "\n",
        "- Files generated for each plasmid\n",
        "\n",
        "<ul>\n",
        "<details><summary>Summary GIF file</summary><hr>\n",
        "<code>plasmid_name.gif</code> file represents a summary overview of the results.\n",
        "The horizontal and vertical axis indicates the position of bases in the original plasmid map and the composition of base calling types: matched, skipped, mismatched, determined to be deletions or insertions, or not_covered.\n",
        "Example of what each type exactly means are shown below. <code>match</code>, <code>mismatch</code>, <code>deletion</code>, and <code>insertion</code> are determined according to the base calling of the read against the plasmid map at the focused position.\n",
        "<code>skipped</code> is applied when there is insertion in other read(s) but not in the read in focus.\n",
        "The term <code>not_covered</code> appears when a read is aligned against the plasmid map but does not cover the whole sequence.\n",
        "\n",
        "<p align=\"center\"><img src=\"https://github.com/MasaakiU/MultiplexNanopore/raw/master/resources/figures/alignment_legend.png\" width=\"1000\"/></p>\n",
        "\n",
        "Because SAVEMONEY is designed assuming that there are no significant differences between the plasmid maps and the actual samples, it is recommended to always check this summary GIF file to confirm that the sample meets that assumption.\n",
        "\n",
        "Below is an example of a typical <code>*.gif</code> file. The preponderance of positions labeled as match or skipped indicates that the plasmid sequence is as expected. The results also indicate that one mutation was found by SAVEMONEY at position 1808, which is displayed with a thin vertical purple line. See <a href=\"#FAQnoGifFileLooksWeird\">Troubleshooting — My GIF file looks weird</a> for more information.\n",
        "\n",
        "<p align=\"center\"><img src=\"https://github.com/MasaakiU/MultiplexNanopore/raw/master/resources/figures/example_gif_1.png\" width=\"1000\"/></p>\n",
        "\n",
        "<hr></details>\n",
        "\n",
        "<details><summary>Consensus FASTQ files</summary><hr>\n",
        "For each plasmid that is detected within the mixture, <a href=\"#Step3\"><b>Step 3. Post-analysis</b></a> will provide two <code>*.fastq</code> files containing consensus sequences (i.e., the sequence of your plasmid) and consensus quality scores.\n",
        "\n",
        "There are two <code>*.fastq</code> files because one of them, <code>plasmid_name.consensus_with_prior.fastq</code>, has used Bayesian analysis to include prior probability of error during plasmid construction, and the other, <code>plasmid_name.consensus_without_prior.fastq</code>, does not include such probabilities. See <a href=\"#Step3no6\"><b>Step 3-6. Calculate consensus</b></a> and <a href=\"https://doi.org/10.1101/2023.04.12.536413\">our publication</a> for more explanation.\n",
        "\n",
        "In most instances, these files contain identical consensus sequences. A manual alignment of them can be performed in your DNA analysis software, e.g., SnapGene, ApE, etc. Below is an example of an alignment of two <code>*.fastq</code> files against the plasmid map using SnapGene. One mutation was detected in the vector backbone.\n",
        "\n",
        "<p align=\"center\"><img src=\"https://github.com/MasaakiU/MultiplexNanopore/raw/master/resources/figures/example_alignment_match.png\" width=\"800\"/></p>\n",
        "\n",
        "If there are no differences between these two <code>*.fastq</code> files, then they represents the sequence of your plasmids and you may stop here.\n",
        "If there are differences at specific positions between these two <code>*.fastq</code> files, we recommend consulting <a href=\"#FAQnoFastqDoesNotMatch\">Troubleshooting — The consensus sequences of the two fastq files do not match</a>, which will guide you to perform <a href=\"#Step4\"><b>Step 4. Visualization of results</b></a>.\n",
        "\n",
        "<hr></details>\n",
        "\n",
        "<details><summary>Alignment consensus (<code>*.ca</code>) files</summary><hr>\n",
        "An alignment consensus file (<code>plasmid_name.ca</code>) contains full alignments information for all reads corresponding to each plasmid. This file can be used to inspect the alignment of all reads to the corresponding plasmid map in <a href=\"#Step4\"><b>Step 4. Visualization of results (optional)</b></a>. This process is roughly the nanopore sequencing equivalent to manually inspecting chromatograms at positions labeled “N” in Sanger sequencing. See <a href=\"#Step4\"><b>Step 4. Visualization of results (optional)</b></a> for more details.\n",
        "<hr></details>\n",
        "</ul>\n",
        "\n",
        "- Files generated for each sample\n",
        "<ul>\n",
        "\n",
        "<details><summary>Summary distribution (<code>*.summary_distribution.svg</code>)</summary><hr>\n",
        "The summary distribution represents a general quality check of the results for each plasmid. Distribution of read lengths and quality scores are displayed in the stacked histogram. The number of rows corresponds to the number of plasmids in the sample, and for each row, the portion that the plasmid occupies in the histogram is shown in blue.\n",
        "\n",
        "<p align=\"center\"><img src=\"https://github.com/MasaakiU/MultiplexNanopore/raw/master/resources/figures/example_summary_distribution.png\" width=\"800\"/></p>\n",
        "\n",
        "<hr></details>\n",
        "\n",
        "<details><summary>Summary scatter (<code>*.summary_scatter.svg</code>)</summary><hr>\n",
        "The summary scatter shows normalized alignment scores for each read against each plasmid map in the sample.\n",
        "In each panel, reads classified to the plasmid shown in the column are indicated with blue dots.\n",
        "\n",
        "Two examples are provided: one with three plasmids that are very different, and one with three plasmids that are more similar. In the latter example, the dots appear closer to the line y=x.\n",
        "\n",
        "The summary scatter can be used to evaluate the success of the classification step and if necessary, modify the `threshold_post` parameter. For example, if the total number of reads was small but each plasmid was sufficiently different (example 1), the threshold can be lowered to ~0.2–0.4 to increase the number of reads that are passed on to the next step of analysis. Conversely, if the total number of reads is large and the plasmids are highly similar to each other (example 2), the quality of the subsequent analysis can be improved by raising the threshold to ~0.6–0.8. See also <a href=\"#Step3no5\"><b>Step 3-5. Set threshold for assignment</b></a>.\n",
        "\n",
        "<p align=\"center\"><img src=\"https://github.com/MasaakiU/MultiplexNanopore/raw/master/resources/figures/example_summary_scatter.png\" width=\"800\"/></p>\n",
        "\n",
        "<hr></details>\n",
        "\n",
        "<details><summary>Summary scatter rotated (<code>*.summary_rotated.svg</code>)</summary><hr>\n",
        "This file contains the same information as <code>*.summary_scatter.svg</code>, but the graph is rotated and expanded horizontally.\n",
        "This transformed version of the summary scatter is particularly beneficial when plasmids that differ by only a few bases were mixed.\n",
        "\n",
        "Example below displays plasmids that differ by only two bases. In the pre-transformation graph on theleft, it is difficult to distinguish if these plasmids were successfully separated or not, but it is clear in the post-transformation graph on the right.\n",
        "\n",
        "<p align=\"center\"><img src=\"https://github.com/MasaakiU/MultiplexNanopore/raw/master/resources/figures/example_summary_scatter_rotated.png\" width=\"800\"/></p>\n",
        "\n",
        "<hr></details>\n",
        "\n",
        "<details><summary>Other files</summary><hr>\n",
        "<ul>\n",
        "<li><code>*.summary_scores.csv</code></li>\n",
        "<dd>Summary score table of normalized alignment score of each read against each plasmid maps. Scores in this file are used to classify reads and to generate summary scatter. Note that alignment scores are not calculated for plasmid map-read pairs with very low similarity to improve processing speed.</dd>\n",
        "<li><code>*.intermediate_results.ir</code></li>\n",
        "<dd>This file can be used to restart the analysis skipping the time-consuming step of <a href=\"#Step3no4\"><b>Step 3-4. Execute alignment</b></a>. Upload this file to the <code>sample_data directory</code> together with your plasmid map files and fastq file(s), and execute the analysis from <a href=\"#Step3no1\"><b>Step 3-2. Click this cell and hit <code>Runtime</code> -> <code>Run after</code></b></a>.</dd>\n",
        "<li><code>*.log</code></li>\n",
        "<dd>Record of settings that generated results.</dd>\n",
        "</ul>\n",
        "<hr></details>\n",
        "\n",
        "</ul>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "oMq3WC9BTBHe"
      },
      "source": [
        "<a name=\"FAQ\"></a>\n",
        "# Troubleshooting\n",
        "\n",
        "Please feel free to create issues at [GitHub issues](https://github.com/MasaakiU/MultiplexNanopore/issues) or contact us ([@MasaakiU5](https://twitter.com/MasaakiU5)) with any questions or feedback.\n",
        "\n",
        "<a name=\"FAQnoGifFileLooksWeird\"></a>\n",
        "<h4><details><summary>My GIF file looks weird</summary><hr>\n",
        "\n",
        "Please check which of the following examples best applies to you and follow the guide.\n",
        "\n",
        "<b>Example 1. Typical result</b><br>\n",
        "The results displayed in this GIF file are acceptable. In the example below, the results indicate one mutation was found by SAVEMONEY at position 1808, which is displayed with a thin vertical purple line. Please examine the two <code>*.fastq</code> files that were generated by SAVEMONEY, following the instructions in <a href=\"#InterpretationOfResultsPost\">Interpretation of results (post-analysis)</a>.\n",
        "\n",
        "<p align=\"center\"><img src=\"https://github.com/MasaakiU/MultiplexNanopore/raw/master/resources/figures/example_gif_1.png\" width=\"1000\"/></p>\n",
        "\n",
        "<b>Example 2. Noisy background</b><br>\n",
        "The results displayed in this GIF file are acceptable. In the example below, although regions of ~300–400 and ~1200–1300 show relatively high noise, this level of noise is not high enough to cause problems for the analysis. Please examine the two <code>*.fastq</code> files that were generated by SAVEMONEY, following the instructions in <a href=\"#InterpretationOfResultsPost\">Interpretation of results (post-analysis)</a>.\n",
        "<p align=\"center\"><img src=\"https://github.com/MasaakiU/MultiplexNanopore/raw/master/resources/figures/example_gif_2.png\" width=\"1000\"/></p>\n",
        "\n",
        "<b>Example 3. Low number of reads</b><br>\n",
        "If the number of reads is small, the GIF file gets even more noisy. Only nine reads were obtained for the example below. Even with such a small number of reads, SAVEMONEY returns a somewhat reliable consensus sequence. Please examine the two <code>*.fastq</code> files that were generated by SAVEMONEY, following the instructions in <a href=\"#InterpretationOfResultsPost\">Interpretation of results (post-analysis)</a>. However, you might sometimes observe discrepancies of consensus sequences between them. In such cases, please follow the instructions in <a href=\"#FAQnoFastqDoesNotMatch\">Troubleshooting — The consensus sequences of the two fastq files do not match</a>. Nanopore sequencing may need to be performed again in some cases.\n",
        "\n",
        "<p align=\"center\"><img src=\"https://github.com/MasaakiU/MultiplexNanopore/raw/master/resources/figures/example_gif_3.png\" width=\"1000\"/></p>\n",
        "\n",
        "<b>Example 4. Local sequence discrepancies between the plasmid map and the plasmid in the sample</b><br>\n",
        "In the example below, a large insertion, indicated by thick green band, was detected in position ~7270. In such a case, SAVEMONEY may not be reporting the correct consensus sequence in the <code>\\*.fastq</code> files for that area in rare occasions. It is recommended to check the raw data around this region according to <a href=\"#Step4\"><b>Step 4 Visualization of results (optional)</b></a>, and then examine the two <code>*.fastq</code> files following the instructions in <a href=\"#InterpretationOfResultsPost\">Interpretation of results (post-analysis)</a>.\n",
        "\n",
        "<p align=\"center\"><img src=\"https://github.com/MasaakiU/MultiplexNanopore/raw/master/resources/figures/example_gif_4.png\" width=\"1000\"/></p>\n",
        "\n",
        "<b>Example 5. No reads at all</b><br>\n",
        "The results indicate that no reads were detected for the plasmid map. There are several possible causes. Check the following factors and perform nanopore sequencing again.\n",
        "<ul>\n",
        "<li>Did your plasmid show the correct band patterns during restriction digestion check?</li>\n",
        "<li>Did you mix equal amounts of plasmids when you made the mixture of plasmids for sample submission?</li>\n",
        "<li>Did the sequencing result in a sufficient number of reads (more than 30 reads per plasmid) for other plasmids in the same sample?</li>\n",
        "<li>Do other plasmids in the sample differ in sequence by two or more bases from the undetected plasmid?</li>\n",
        "<li>Is your plasmid map correct, including vector backbone? If you have mixed plasmids with similar sequences, and the vector backbone sequence differs between the map and the actual plasmid, reliable analysis may not be performed.</li>\n",
        "</ul>\n",
        "<p align=\"center\"><img src=\"https://github.com/MasaakiU/MultiplexNanopore/raw/master/resources/figures/example_gif_5.png\" width=\"1000\"/></p>\n",
        "\n",
        "<hr></details>\n",
        "\n",
        "<a name=\"FAQnoFastqDoesNotMatch\"></a>\n",
        "<h4><details><summary>The consensus sequences of the two FASTQ files do not match</summary><hr>\n",
        "\n",
        "If there are differences at specific positions between the two <code>*.fastq</code> files, your data quality might not be good enough to generate reliable consensus sequences, most likely due to the low number of reads. Usually in this instance, <code>plasmid_name.consensus_with_prior.fastq</code> tells you that sequences are the same as the reference, and <code>plasmid_name.consensus_without_prior.fastq</code> tells you that there is a mutation.\n",
        "\n",
        "<p align=\"center\"><img src=\"https://github.com/MasaakiU/MultiplexNanopore/raw/master/resources/figures/example_alignment_mismatch.png\" width=\"600\"/></p>\n",
        "\n",
        "In this case, it is not fully rigorous to claim the absolute correctness of either the “with prior” results indicating that there is an error or the “without prior” results indicating that there is no error. In other words, there is a good chance that the sequence of your plasmid is correct. Do not discard your plasmid!\n",
        "We recommend taking the following steps to examine your plasmid sequence:\n",
        "<ol>\n",
        "<li>\n",
        "Inspect the alignment of all reads corresponding to the position of this plasmid, following <a href=\"#Step4\"><b>Step 4. Visualization of results (optional)</b></a>. This process is roughly the nanopore sequencing equivalent to manually inspecting chromatograms at positions labeled “N” in Sanger sequencing. To do so, you must first perform the alignment visualization below. </li><li>\n",
        "If the number of reads is low (less than 30), this may be the reason for the inability to determine the consensus base calling.\n",
        "If you cannot make a consensus base call with confidence even with the help of <a href=\"#Step4\"><b>Step 4. Visualization of results (optional)</b></a>, we recommend double-checking your plasmid concentration and quality, preparing new samples, and performing new nanopore sequencing.\n",
        "</li><li>\n",
        "If the number of reads is high enough (more than 30) but you still see discrepancy, your plasmid may contain multiple sequences. We recommend purifying the plasmid again from a single colony and send repeating nanopore sequencing.\n",
        "</li>\n",
        "</ol>\n",
        "<hr></details>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "<a name=\"ReleaseNotes\"></a>\n",
        "# Release notes\n",
        "\n",
        "<a name=\"ver0no2no11\"></a>\n",
        "<h4><details><summary>ver_0.2.12 (this version)</summary><hr>\n",
        "<h3>Internal changes</h3>\n",
        "<ul>\n",
        "    <li>Updates in the msa parameter (1 > 5, -2 > -4, -3 > -8, -1 > -6 for <code>m</code>, <code>n</code>, <code>g</code>, <code>e</code>, respectively)\n",
        "    <li>Minor updates in display style of alignment in Step 4.\n",
        "</ul>\n",
        "<hr></details>\n",
        "\n",
        "<a name=\"ver0no2no11\"></a>\n",
        "<h4><details><summary><a href=\"https://colab.research.google.com/github/MasaakiU/MultiplexNanopore/blob/master/colab/MultiplexNanopore_ver_0.2.11.ipynb\">ver_0.2.11</a></summary><hr>\n",
        "\n",
        "<h3>New features</h3>\n",
        "<ul>\n",
        "    <li>Step 1. Pre-survey\n",
        "        <ul>\n",
        "            <li>Faster execution speed\n",
        "            <li>Support for setting <code>number_of_groups</code>\n",
        "            <li>Accepts a large number (~100) of plasmids\n",
        "        </ul>\n",
        "    <li>Step 3. Post-analysis\n",
        "        <ul>\n",
        "            <li>Faster execution speed\n",
        "            <li>Improved estimation of the sequence when wrong plasmid maps are provided\n",
        "            <li>Reduced size of exported files\n",
        "        </ul>\n",
        "    <li>Step 4. Visualization of results\n",
        "        <ul>\n",
        "            <li>Improved display style of alignment results\n",
        "            <li>Accepts <code>offset</code> parameter when displaying alignment results\n",
        "        </ul>\n",
        "    </ul>\n",
        "<h3>Internal changes</h3>\n",
        "<ul>\n",
        "    <li>The parameter <code>threshold_pre</code> was renamed to <code>distance_threshold</code>\n",
        "    <li>The parameter <code>threshold_post</code> was renamed to <code>score_threshold</code>\n",
        "    <li>Format of output files were changed\n",
        "        <ul>\n",
        "            <li><code>recommended_grouping.txt</code> (<- <code>recommended_groupings.txt</code>)\n",
        "            <li><code>recommended_grouping.pdf</code> (<- <code>recommended_group_#.svg</code>)\n",
        "            <li><code>fastq_file_name.log</code> (<- <code>log_with/without_prior.txt</code>)\n",
        "            <li><code>fastq_file_name.csv</code> (<- <code>summary_scores.txt</code>)\n",
        "            <li><code>fastq_file_name.summary_rotated.svg</code> (new)\n",
        "            <li><code>fastq_file_name.intermediate_results.ir</code> (<- <code>fastq_file_name.intermediate_results.ir</code>)\n",
        "            <li><code>plasmid_name.fa.ca</code> (<- <code>plasmid_name.alignment_with/without_prior.txt</code>)\n",
        "        </ul>\n",
        "    <li>All codes were replaced with <a href=\"https://pypi.org/project/savemoney/\">PyPI</a>-based scripts\n",
        "</ul>\n",
        "<hr></details>\n",
        "\n",
        "<a name=\"ver0no1no0\"></a>\n",
        "<h4><details><summary><a href=\"https://colab.research.google.com/github/MasaakiU/MultiplexNanopore/blob/master/colab/MultiplexNanopore_ver_0.1.0.ipynb\">ver_0.1.0 (original)</a></summary><hr>\n",
        "\n",
        "Basic piplein of SAVEMONEY was created.\n",
        "<ul>\n",
        "    <li>Step 1. Pre-survey\n",
        "        <ul>\n",
        "            <li>Step 1-1. Upload plasmid maps\n",
        "            <li>Step 1-2. Advanced settings\n",
        "            <li>Step 1-3. Click this cell and hit <code>Runtime</code> -> <code>Run the focused cell</code>\n",
        "            <li>Step 1-4. View pre-survey results\n",
        "        </ul>\n",
        "    <li>Step 2. Submit samples\n",
        "    <li>Step 3. Post-analysis\n",
        "        <ul>\n",
        "            <li>Step 3-1. Upload files\n",
        "            <li>Step 3-2. Click this cell and hit <code>Runtime</code> -> <code>Run after</code>\n",
        "            <li>Step 3-3. Select files\n",
        "            <li>Step 3-4. Execute alignment\n",
        "            <li>Step 3-5. Set threshold for assignment\n",
        "            <li>Step 3-6. Calculate consensus\n",
        "            <li>Step 3-7. Export results\n",
        "        </ul>\n",
        "    <li>Step 4. Visualization of results (optional)\n",
        "        <ul>\n",
        "            <li>Step 4-1. Upload files\n",
        "            <li>Step 4-2. Set parameters\n",
        "            <li>Step 4-3. Select this cell and hit <code>Runtime</code> -> <code>Run after</code>\n",
        "        </ul>\n",
        "</ul>\n",
        "<hr></details>\n",
        "\n"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3.10.4 64-bit ('env_RAPID')",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "name": "python",
      "version": "3.10.4"
    },
    "vscode": {
      "interpreter": {
        "hash": "65fb75ef547f7f70fb8dbc6edf61adc9b4cedb2ccbdc90e1548a5e817107408e"
      }
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
